### ISSUE TITLE
Render Rounded titlebars

### ISSUE THREAD

**Issue by WillPower3309** [WillPower3309/swayfx#19]
- **Description:**
When window corners are rounded, titlebars need to be supported too!

**Comment by ErikReider** [comment]
I'm not at all experienced with GLSL but maybe this is a solution?

![image](https://user-images.githubusercontent.com/35975961/187037137-b94b8e44-b353-4585-95a1-dc23e0abc21f.png)
![image](https://user-images.githubusercontent.com/35975961/187037140-be264484-73de-43fb-9b0d-9b1a00d75c56.png)

Shrink the borders and draw the rounded corners separately?

I'm not sure about how to deal with the dual colored titlebar, maybe clipping but that wouldn't look that good with multiple colors...
![image](https://user-images.githubusercontent.com/35975961/187037357-63eb0bfb-af4c-4a9a-acb5-7067721e5136.png)

**Comment by WillPower3309** [comment]
Good idea! I think I'll go this route for rendering the borders, (that is, render them as usual but make them smaller to avoid the corners) and then separately render rounded corner borders on the 4

**Comment by WillPower3309** [comment]
> I'm not at all experienced with GLSL but maybe this is a solution?
> 
> ![image](https://user-images.githubusercontent.com/35975961/187037137-b94b8e44-b353-4585-95a1-dc23e0abc21f.png)
> ![image](https://user-images.githubusercontent.com/35975961/187037140-be264484-73de-43fb-9b0d-9b1a00d75c56.png)
> 
> Shrink the borders and draw the rounded corners separately?
> 
> I'm not sure about how to deal with the dual colored titlebar, maybe clipping but that wouldn't look that good with multiple colors...
> ![image](https://user-images.githubusercontent.com/35975961/187037357-63eb0bfb-af4c-4a9a-acb5-7067721e5136.png)
> 

For titlebars, I was thinking I could render the top left and top right corners of the window as sharp corners, then round the top titlebar corners

**Comment by WillPower3309** [comment]
Alternatively, we could have titlebars render as follows:
https://i.redd.it/2d6lpyq8t8l91.png

**Comment by ErikReider** [comment]
Wow! That looks a lot better!

**Comment by RicArch97** [comment]
> Alternatively, we could have titlebars render as follows: https://i.redd.it/2d6lpyq8t8l91.png

Perhaps in this case it would also be nice to have to option to only round the outer border but not the inner

**Comment by WillPower3309** [comment]
> Alternatively, we could have titlebars render as follows: https://i.redd.it/2d6lpyq8t8l91.png

The top corners of the window past the radius can blend with the titlebar color by updating the tex_frag shaders to something like this:
```glsl
"void main() {\n"
"   gl_FragColor = texture2D(tex, v_texcoord) * alpha;\n"
"   vec2 corner_distance = min(gl_FragCoord.xy - position, position + vec2(width, height) - gl_FragCoord.xy);\n"
"   if (max(corner_distance.x, corner_distance.y) < radius) {\n"
"		float d = radius - distance(corner_distance, vec2(radius, radius));\n"
"		float smooth = smoothstep(-1.0f, 1.0f, d);\n"
"		gl_FragColor = mix(TITLEBAR_COLOR, gl_FragColor, smooth);\n"
"   }\n"
"}\n";
```

**Comment by WillPower3309** [comment]
> > Alternatively, we could have titlebars render as follows: https://i.redd.it/2d6lpyq8t8l91.png
> 
> Perhaps in this case it would also be nice to have to option to only round the outer border but not the inner, though you would need more space at the bottom

This will happen automatically depending on how thick the window border is in relation to the corner radius once #24 is merged
